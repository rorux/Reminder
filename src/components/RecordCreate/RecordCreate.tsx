import React, { useCallback, useEffect } from 'react';
import Grid from '@mui/material/Grid';
import RecordSelect from '@components/formsUI/RecordSelect';
import Typography from '@mui/material/Typography';
import TextField from '@mui/material/TextField';
import LoadingButton from '@mui/lab/LoadingButton';
import SendIcon from '@mui/icons-material/Send';
import {
  weeklyComputedDays,
  monthlyComputedDays,
  quarterlyComputedDays,
} from '@utils/records/ComputedDays';
import { TPeriod, TWeekDay, THolidays, TMonthDays, TQuarter } from '@utils/records/types';
import { PERIOD, WEEKDAY, HOLIDAYS, MONTH_DAYS, QUARTER } from '@utils/records/constants';

const RecordCreate = () => {
  const [title, setTitle] = React.useState('');
  const [periodValue, setPeriodValue] = React.useState<TPeriod>('');

  const [weeklyWeekDay, setWeeklyWeekDay] = React.useState<TWeekDay>('');
  const [weeklyHolidays, setWeeklyHolidays] = React.useState<THolidays>('');

  const [monthlyDay, setMonthlyDay] = React.useState<TMonthDays>('');
  const [monthlyHolidays, setMonthlyHolidays] = React.useState<THolidays>('');

  const [quarterlyMonths, setQuarterlyMonths] = React.useState<TQuarter>('');
  const [quarterlyDay, setQuarterlyDay] = React.useState<TMonthDays>('');
  const [quarterlyHolidays, setQuarterlyHolidays] = React.useState<THolidays>('');

  const [isFormReady, setIsFormReady] = React.useState<TPeriod>('');
  const [loadingSubmit, setLoadingSubmit] = React.useState<boolean>(false);

  useEffect(() => {
    if (periodValue === 'weekly' && title && weeklyWeekDay && weeklyHolidays) {
      setIsFormReady('weekly');
    } else if (periodValue === 'monthly' && title && monthlyDay && monthlyHolidays) {
      setIsFormReady('monthly');
    } else if (
      periodValue === 'quarterly' &&
      title &&
      quarterlyMonths &&
      quarterlyDay &&
      quarterlyHolidays
    ) {
      setIsFormReady('quarterly');
    } else {
      setIsFormReady('');
    }
  }, [
    title,
    periodValue,
    weeklyWeekDay,
    weeklyHolidays,
    monthlyDay,
    monthlyHolidays,
    quarterlyMonths,
    quarterlyDay,
    quarterlyHolidays,
  ]);

  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setTitle(event.target.value as string);
  };

  const handleClickSubmit = useCallback(() => {
    setLoadingSubmit(true);
    let days: Array<number> = [];
    if (periodValue === 'weekly') {
      days = weeklyComputedDays(weeklyWeekDay, weeklyHolidays);
    } else if (periodValue === 'monthly') {
      days = monthlyComputedDays(monthlyDay, monthlyHolidays);
    } else if (periodValue === 'quarterly') {
      days = quarterlyComputedDays(quarterlyMonths, quarterlyDay, quarterlyHolidays);
    } else throw Error;
    console.log(days);
    setLoadingSubmit(false);
  }, [
    periodValue,
    weeklyWeekDay,
    weeklyHolidays,
    monthlyDay,
    monthlyHolidays,
    quarterlyMonths,
    quarterlyDay,
    quarterlyHolidays,
  ]);

  return (
    <div data-test="record-create">
      <Typography sx={{ mb: 1 }} variant="overline" display="block" gutterBottom>
        Заполните форму
      </Typography>
      <Grid container spacing={3} maxWidth={600}>
        <Grid item xs={12}>
          <TextField
            value={title}
            onChange={handleTitleChange}
            label="Наименование"
            variant="outlined"
            fullWidth
            required
          />
        </Grid>
        <Grid item xs={12}>
          <RecordSelect name="Периодичность" params={PERIOD} setter={setPeriodValue} />
        </Grid>
        {periodValue === 'weekly' && (
          <>
            <Grid item xs={12}>
              <RecordSelect name="День недели" params={WEEKDAY} setter={setWeeklyWeekDay} />
            </Grid>
            <Grid item xs={12}>
              <RecordSelect
                name="Праздники и выходные"
                params={HOLIDAYS}
                setter={setWeeklyHolidays}
              />
            </Grid>
          </>
        )}
        {periodValue === 'monthly' && (
          <>
            <Grid item xs={12}>
              <RecordSelect name="Число" params={MONTH_DAYS} setter={setMonthlyDay} />
            </Grid>
            <Grid item xs={12}>
              <RecordSelect
                name="Праздники и выходные"
                params={HOLIDAYS}
                setter={setMonthlyHolidays}
              />
            </Grid>
          </>
        )}
        {periodValue === 'quarterly' && (
          <>
            <Grid item xs={12}>
              <RecordSelect name="Месяцы" params={QUARTER} setter={setQuarterlyMonths} />
            </Grid>
            <Grid item xs={12}>
              <RecordSelect name="Число" params={MONTH_DAYS} setter={setQuarterlyDay} />
            </Grid>
            <Grid item xs={12}>
              <RecordSelect
                name="Праздники и выходные"
                params={HOLIDAYS}
                setter={setQuarterlyHolidays}
              />
            </Grid>
          </>
        )}
        <Grid item>
          <LoadingButton
            variant="contained"
            color="success"
            disabled={!isFormReady}
            endIcon={<SendIcon />}
            loadingPosition="end"
            loading={loadingSubmit}
            onClick={handleClickSubmit}
            size="large"
          >
            Сохранить
          </LoadingButton>
        </Grid>
      </Grid>
    </div>
  );
};

export default RecordCreate;
