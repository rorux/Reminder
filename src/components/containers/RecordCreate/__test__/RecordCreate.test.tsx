import React from 'react';
import * as Redux from 'react-redux';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import LoadingButton from '@mui/lab/LoadingButton';
import RecordCreate from '../RecordCreate';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('@utils/records/ComputedDays');

const {
  weeklyComputedDays,
  monthlyComputedDays,
  quarterlyComputedDays,
} = require('@utils/records/ComputedDays');

describe('RecordCreate component', () => {
  let component: Enzyme.ShallowWrapper;
  let useDispatch: any;
  let mockDispatch: jest.Mock;

  beforeEach(() => {
    useDispatch = jest.spyOn(Redux, 'useDispatch');
    mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    // setState = jest.fn();
    // useStateSpy = jest.spyOn(React, 'useState');
    // // @ts-ignore
    // useStateSpy.mockImplementation((init) => [init, setState]);
    component = shallow(<RecordCreate />);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('Should render RecordCreate component', () => {
    expect(component).toMatchSnapshot();
  });
  it('Should call weeklyComputedDays when click Submit', () => {
    const selectPeriodValue = component.find(`[data-test='select-period-value']`);
    // @ts-ignore
    selectPeriodValue.props().setter('weekly');
    // @ts-ignore
    component.find(LoadingButton).props().onClick();
    expect(weeklyComputedDays).toHaveBeenCalled();
  });

  it('Should call monthlyComputedDays when click Submit', () => {
    const selectPeriodValue = component.find(`[data-test='select-period-value']`);
    // @ts-ignore
    selectPeriodValue.props().setter('monthly');
    // @ts-ignore
    component.find(LoadingButton).props().onClick();
    expect(monthlyComputedDays).toHaveBeenCalled();
  });

  it('Should call quarterlyComputedDays when click Submit', () => {
    const selectPeriodValue = component.find(`[data-test='select-period-value']`);
    // @ts-ignore
    selectPeriodValue.props().setter('quarterly');
    // @ts-ignore
    component.find(LoadingButton).props().onClick();
    expect(quarterlyComputedDays).toHaveBeenCalled();
  });
});
