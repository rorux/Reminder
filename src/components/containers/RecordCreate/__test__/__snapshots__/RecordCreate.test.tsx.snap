// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordCreate component Should render RecordCreate component 1`] = `
<div
  data-test="record-create"
>
  <ForwardRef(Typography)
    display="block"
    gutterBottom={true}
    sx={
      Object {
        "mb": 1,
      }
    }
    variant="overline"
  >
    Заполните форму
  </ForwardRef(Typography)>
  <ForwardRef(Grid)
    container={true}
    maxWidth={600}
    spacing={3}
  >
    <ForwardRef(Grid)
      item={true}
      xs={12}
    >
      <ForwardRef(TextField)
        data-test="input-title"
        fullWidth={true}
        label="Наименование"
        onChange={[Function]}
        required={true}
        value=""
        variant="outlined"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      item={true}
      xs={12}
    >
      <RecordSelect
        data-test="select-period-value"
        name="Периодичность"
        params={
          Array [
            Object {
              "name": "Еженедельно",
              "value": "weekly",
            },
            Object {
              "name": "Ежемесячно",
              "value": "monthly",
            },
            Object {
              "name": "Ежеквартально",
              "value": "quarterly",
            },
          ]
        }
        setter={[Function]}
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      item={true}
    >
      <ForwardRef(LoadingButton)
        color="success"
        disabled={true}
        endIcon={<Memo />}
        loading={false}
        loadingPosition="end"
        onClick={[Function]}
        size="large"
        variant="contained"
      >
        Сохранить
      </ForwardRef(LoadingButton)>
    </ForwardRef(Grid)>
  </ForwardRef(Grid)>
</div>
`;
